### PR TITLE
[Score API] Add Multi-Item Scoring with pre-computed delimiter indices

### PR DESCRIPTION
**Pull request by fortunecookiee**
## Motivation

Multi-Item Scoring (MIS) packs a query and multiple items into a single sequence (`query<delim>item1<delim>item2<delim>...`) for a single forward pass instead of N separate passes. The existing `--multi-item-scoring-delimiter` required callers to choose and pass a delimiter token ID, and the runtime scanned `input_ids` on the GPU to find delimiter positions at every stage (attention, logits, pooler), causing redundant work and implicit CUDA syncs from `.item()` and `.nonzero()` calls.

This PR replaces `--multi-item-scoring-delimiter <token_id>` with a simple `--enable-mis` boolean flag and pre-computes delimiter indices from item lengths, eliminating all runtime scanning and GPU-to-CPU syncs.

## `--enable-mis` flag

**Before** — callers had to pick a delimiter token ID and pass incompatible-setting flags manually:
```bash
--multi-item-scoring-delimiter 128009 --disable-radix-cache --chunked-prefill-size -1
```

**After** — a single boolean flag; required settings are auto-applied:
```bash
--enable-mis --attention-backend flashinfer
```

The flag:
- Hardcodes `MIS_DELIMITER_TOKEN_ID = 9999` as a placeholder token (exists only for FlashInfer attention mask compatibility and logprob column indexing — positions are determined by item lengths, not by scanning for this token)
- **Auto-disables** CUDA graph, radix cache, and chunked prefill (with warnings)
- **Asserts** `--attention-backend flashinfer` (not auto-changed since the backend choice is user-facing)
- Removes the need for callers to know tokenizer-specific delimiter tokens

All MIS setup and validation is consolidated in `_handle_multi_item_scoring()`, which runs after other `_handle_*` helpers so the final config state is authoritative.

## Pre-computed delimiter indices

Instead of scanning `input_ids == delimiter` on the GPU at each stage, delimiter positions are computed once during tokenization from item lengths and piped through the request as CPU tensors:

```
tokenizer_manager_score_mixin  →  io_struct  →  scheduler  →  schedule_batch (CPU tensor)
    →  forward_batch_info  →  flashinfer_backend / logits_processor / pooler
```

The ScheduleBatch enforces batch consistency: if any request in a batch carries MIS indices, all must (asserted) — mixed MIS/non-MIS batches are rejected early with a clear error.

### Consumers updated
- **`flashinfer_backend.py`**: `_process_multi_item_scoring()` uses pre-computed CPU tensors + `torch.searchsorted` instead of `(input_ids == delim).nonzero()`. Uses `extend_prefix_lens_cpu` / `extend_seq_lens_cpu` to avoid GPU syncs.
- **`logits_processor.py`**: `compute_logprobs_for_multi_item_scoring()` uses pre-computed indices to gather hidden states at delimiter positions. Builds offset-adjusted indices on CPU and does a single batched `.to(device)` transfer instead of per-request transfers.
- **`pooler.py`**: new `pool_at_delimiter_positions()` helper replaces inline `input_ids` scanning. `score_and_pool()` now pools hidden states at pre-delimiter positions first, then scores only those tokens — preserves upstream's per-request `pooled_hidden_states` return feature while using pre-computed indices.
- **`scheduler_output_processor_mixin.py`**: `_is_multi_item_scoring()` uses `enable_mis` + pre-computed indices instead of scanning `origin_input_ids`.

## Tests
- New `test/registered/prefill_only/test_multi_item_scoring.py` with CI registration — 6 test classes: ServerArgs validation, generation MIS, classification MIS (pre-trained model), parity, classification parity, and advanced (distinctness, determinism, concurrency).
- Updated `test_score_api.py`, `test_score_engine.py`, `test_embed_overrides.py`, `test_pooled_hidden_states.py` to use `--enable-mis` / `enable_mis=True`.
- Unit test `test_pooler_score_and_pool.py` exercises `pool_at_delimiter_positions` and the MIS path of `score_and_pool` on CPU — no GPU required.

Co-authored-by: Chanh Nguyen <chanhnguyen@gmail.com>
Co-authored-by: Sundara Raman Ramachandran <sundar24295@gmail.com>

### CODE REVIEW

File: python/sglang/srt/server_args.py [hunk 1]
```diff
@@ -168,6 +167,18 @@
...
+# Format: query<delim>item1<delim>item2<delim>...
+#
+# This token is a structural placeholder only — delimiter *positions* are pre-computed
+# from item lengths (multi_item_delimiter_indices), not by scanning for this token ID.
+# The token must exist in the sequence so that:
+#   1. FlashInfer's attention mask can identify item boundaries,
+#   2. logprob extraction has a valid column index to read from.
+# This coupling to FlashInfer's signature is temporary and will be removed once
+# the attention backend supports position-only MIS without a sentinel token.
+MIS_DELIMITER_TOKEN_ID = 9999
```
Comment (kpham-sgl) [inline python/sglang/srt/server_args.py:167]: 1. What happen when model vocab size > 9999 (and token_id is a meaningful token in the model vocab) 
2. What happen when model vocab size < 9999

File: python/sglang/srt/layers/logits_processor.py [hunk 2]
```diff
@@ -1090,19 +1082,22 @@ def compute_logprobs_for_multi_item_scoring(
...
-            next_token_logits=None,  # Multi-item scoring doesn't need next token logits
+            next_token_logits=None,
             input_token_logprobs=input_token_logprobs,
             input_top_logprobs_val=input_top_logprobs_val,
             input_top_logprobs_idx=input_top_logprobs_idx,
             input_token_ids_logprobs_val=input_token_ids_logprobs_val,
             input_token_ids_logprobs_idx=input_token_ids_logprobs_idx,
-            # FIXME: These fields are not logits-related but are passed through here as a
-            # workaround since ForwardBatch is local to forward_batch_generation().
-            # They should be moved to GenerationBatchResult to keep this class clean.
```
Comment (kpham-sgl) [inline python/sglang/srt/layers/logits_processor.py:1105]: This comment should be left intact

File: python/sglang/srt/layers/logits_processor.py [hunk 3]
```diff
@@ -1006,39 +1007,44 @@ def compute_logprobs_for_multi_item_scoring(
...
+                logits_metadata.extend_seq_lens_cpu, multi_item_delimiter_indices
+            ):
+                if len(indices_tensor) > 0:
+                    # Note: if the first delimiter is at position 0 (empty query),
+                    # indices - 1 wraps to -1. This is harmless — the first
+                    # delimiter entry is always discarded by
+                    # _process_multi_item_scoring_results.
+                    all_tensors.append(
+                        indices_tensor.to(device, non_blocking=True) + offset - 1
+                    )
```
Comment (kpham-sgl) [inline python/sglang/srt/layers/logits_processor.py]: Perf nit: I believe the `indices_tensor.to()` here is still implicitly blocking because `indices_tensor` is non-pinned memory. If you *really* want to speedup the performance I suggest batching all of these `.to(device)` call to GPU. You can still later check `len(indices_tensor) > 0` without it materialize on CPU. Disregard this comment if current behaviour is intended / acceptable

Further improvement for this snippet (including making it CUDA graphable is to allocate fixed size `multi_item_indices` and write even for len(indices_tensor) > 0 case.

File: python/sglang/srt/layers/pooler.py [hunk 4]
```diff
@@ -26,52 +25,68 @@ class EmbeddingPoolerOutput:
...
+    for req_idx, req_seq_len in enumerate(forward_batch.extend_seq_lens_cpu):
+        indices_tensor = forward_batch.multi_item_delimiter_indices[req_idx]
+        n = len(indices_tensor)
+        if n > 0:
+            # Note: if the first delimiter is at position 0 (empty query),
+            # indices - 1 wraps to -1. This is harmless — the first delimiter
+            # entry is always discarded by _process_multi_item_scoring_results.
+            all_index_tensors.append(
+                indices_tensor.to(device, non_blocking=True) + offset - 1
+            )
```
Comment (kpham-sgl) [inline python/sglang/srt/layers/pooler.py]: ditto

File: python/sglang/srt/layers/pooler.py [hunk 5]
```diff
@@ -26,52 +25,68 @@ class EmbeddingPoolerOutput:
...
-    # Standard classification path: score all tokens, then pool.
-    logits = score_head(hidden_states)
-    pooled_logits = pooler(logits, forward_batch).embeddings
-    return EmbeddingPoolerOutput(embeddings=pooled_logits)
+    if forward_batch.multi_item_delimiter_indices is not None:
+        logits = score_head(hidden_states)
+        scores = pool_at_delimiter_positions(logits, forward_batch, input_ids.device)
+    else:
+        pooled_hidden = pooler(hidden_states, forward_batch).embeddings
+        scores = score_head(pooled_hidden)
```
Comment (kpham-sgl) [inline python/sglang/srt/layers/pooler.py]: IIUC here we use `pool_at_delimiter_positions` when it is MIS and "normal" pooler for single item. scores` type seem to differ in the two code path: for MIS, this is a list of tensor, for single item its a single tensor. 

Any concern for downstream consumer introduced in https://github.com/sgl-project/sglang/pull/22427 because of this differences?

File: python/sglang/srt/managers/schedule_batch.py [hunk 6]
```diff
@@ -1803,6 +1808,21 @@ def prepare_for_extend(self):
...
+            r.multi_item_delimiter_indices is not None for r in reqs
+        ):
+            self.multi_item_delimiter_indices = [
+                (
+                    torch.tensor(r.multi_item_delimiter_indices, dtype=torch.int64)
+                    if r.multi_item_delimiter_indices is not None
+                    else None
+                )
+                for r in reqs
+            ]
```
Comment (kpham-sgl) [inline python/sglang/srt/managers/schedule_batch.py:1838]: Consumer of `self.multi_item_delimiter_indices` always a non-None element in the list. I assume once MIS is turned on the server does not expect a non MIS request?

File: python/sglang/srt/managers/tokenizer_manager_score_mixin.py [hunk 7]
```diff
@@ -592,33 +587,38 @@ async def score_request(
...
+            _, input_ids, positional_embed_overrides, delimiter_indices = (
+                self._build_token_id_inputs(
+                    query_ids,
+                    items_ids,
+                    item_first,
+                    use_multi_item_scoring,
+                    embed_override_token_id,
+                    query_embed_overrides,
+                    item_embed_overrides,
+                )
```
Comment (kpham-sgl) [inline python/sglang/srt/managers/tokenizer_manager_score_mixin.py:638]: ditto

File: python/sglang/srt/managers/tokenizer_manager_score_mixin.py [hunk 8]
```diff
@@ -592,33 +587,38 @@ async def score_request(
...
```
Comment (kpham-sgl) [inline python/sglang/srt/managers/tokenizer_manager_score_mixin.py:652]: ditto

File: python/sglang/srt/managers/tokenizer_manager_score_mixin.py [hunk 9]
```diff
@@ -550,31 +545,31 @@ async def score_request(
...
+                delimiter_token_id = MIS_DELIMITER_TOKEN_ID
                 query_ids, items_ids = self._batch_tokenize_query_and_items(
                     query, items_list
                 )
-                combined_input_ids = self._build_multi_item_token_sequence(
-                    query_ids, items_ids, delimiter_token_id
+                combined_input_ids, delimiter_indices = (
+                    self._build_multi_item_token_sequence(
+                        query_ids, items_ids, delimiter_token_id
+                    )
```
Comment (kpham-sgl) [inline python/sglang/srt/managers/tokenizer_manager_score_mixin.py:611]: why wrap this in ()? Linting issue?

**Review by kpham-sgl — COMMENTED** [review]
@fortunecookiee I left some comments (mostly small). Three bigger questions 
- What is the relationship of this PR and https://github.com/sgl-project/sglang/pull/22427? Are they independent or would one (slightly) change the behavior of the other?
- Can you profile to verify that this PR really brings improvement? 
(remove all unnecessary syncs, etc)
- Is there a plan to support Cudagraph for MIS? I think many code snippets in this PR can be make into Cudagraph compatible but require careful thinking.

File: python/sglang/srt/server_args.py (hunk 1 above)
Comment (fortunecookiee) [reply python/sglang/srt/server_args.py:167]: Delimiter positions are precomputed from item lengths — nothing scans for token 9999.

The pooler extracts logits at delimiter_position - 1 (the last token of each item), so the delimiter’s own output is never used for scoring.

The custom FlashInfer kernel does not compute the delimiter’s embedding during the attention computation. It only uses the delimiter positions to identify item boundaries for masking.

However, the token is still indexed from the model’s embedding vocabulary table during embed_tokens. If vocab_size < 9999, this will throw an index out of bounds error.

In practice, this does not happen because LLM vocabulary sizes are typically much larger than 9999 (for example, Llama: 128K, Qwen: 152K).

File: python/sglang/srt/layers/logits_processor.py (hunk 2 above)
Comment (fortunecookiee) [reply python/sglang/srt/layers/logits_processor.py:1105]: restored.

File: python/sglang/srt/layers/logits_processor.py (hunk 3 above)
Comment (fortunecookiee) [reply python/sglang/srt/layers/logits_processor.py]: Thanks for the suggestions! I updated the code to stack CPU tensors and perform a single CPU–GPU transfer. The current non_blocking setting is effectively a no-op. I also experimented with pin_memory, but it caused a noticeable performance regression for smaller tensors, so I’ll keep the current approach for now.

For CUDA Graph, yes—we will address scoring with MIS by providing end-to-end support using piecewise CUDA Graphs

File: python/sglang/srt/layers/pooler.py (hunk 5 above)
Comment (fortunecookiee) [reply python/sglang/srt/layers/pooler.py]: Yeah, I’m working on addressing the current comments first and will rebase onto the head of master since #22427 has been merged.

File: python/sglang/srt/managers/schedule_batch.py (hunk 6 above)
Comment (fortunecookiee) [reply python/sglang/srt/managers/schedule_batch.py:1838]: yes, dropped the `None` item

File: python/sglang/srt/managers/tokenizer_manager_score_mixin.py (hunk 9 above)
Comment (fortunecookiee) [reply python/sglang/srt/managers/tokenizer_manager_score_mixin.py:611]: yes, added by auto-formatter

Comment (kpham-sgl) [reply python/sglang/srt/managers/tokenizer_manager_score_mixin.py:611]: Hmm by our pre-commit?

Comment (fortunecookiee) [reply python/sglang/srt/managers/tokenizer_manager_score_mixin.py:611]: it looks needed to avoid too long function call breaking num of characters limits
```
So yes, the outer parens ARE required for this specific line due to how deep the indentation is. Black mathematically forces the choice based on line length at that indentation level.

  If you wanted to remove them, the only way would be to rename the variables (e.g., drop combined_input_ids to combined_ids) or rename the method (_build_multi_item_token_sequence → shorter). Not worth it for a
  style preference.

  Conclusion: keep as-is — it's what black produces.
```

**Comment by fortunecookiee** [comment]
@kpham-sgl for the additional parens
```
All 6 touched files pass black's check as-is. This means every added = ( is what black produced — it's either required by line-length at that indent depth, or it wraps a ternary/nested expression for clarity.

  Examples from the diff:

  1. forward_batch.positions[seq_start + first_delim : seq_end] = ( — assignment target is long (47 chars at indent 12), total line would be ~100+ chars without wrap.
  2. multi_item_delimiter_indices=( — this is the common pattern:
  multi_item_delimiter_indices=(
      self.multi_item_delimiter_indices[i]
      if self.multi_item_delimiter_indices is not None
      else None
  )
  2. Ternary expression wrapping — black always wraps multi-line conditionals in parens.
  3. combined_input_ids, delimiter_indices = ( — tuple-unpacking from long call, line-length hits 88+ at deep indent.
  4. positional_embed_overrides = ( — my own explicit refactor to extract the ternary; wrapped for readability.
  5. _, input_ids, positional_embed_overrides, delimiter_indices = ( — 4-element unpacking + long method call; the flat form would be well over 88 chars even at the outermost indent.

  Conclusion: All outer parens in the PR are black-formatter output. None can be removed without either:
  - Making lines exceed 88 chars (black would add them back), or
  - Hurting readability (ternary expressions split across lines need parens for clarity).

  The current formatting is what black produces — it's the SGLang project's enforced style.
```

**Comment by fortunecookiee** [comment]
> @fortunecookiee I left some comments (mostly small). Three bigger questions
> 
> * What is the relationship of this PR and #22427? Are they independent or would one (slightly) change the behavior of the other?
> * Can you profile to verify that this PR really brings improvement?
>   (remove all unnecessary syncs, etc)
> * Is there a plan to support Cudagraph for MIS? I think many code snippets in this PR can be make into Cudagraph compatible but require careful thinking.

- rebased on #22427
- our internal benchmark shows ~4.5% throughput gain by removing the mem_syncs
Metric | Before | After | Delta
-- | -- | -- | --
Throughput (RPS) | 152.6 | 159.4 | +4.5%
Items/sec | 7,628.2 | 7,971.3 | +4.5%
- Currently, if enable-mis is set, CUDA Graph is automatically disabled. We’re working on enabling score + MIS to run with piecewise CUDA Graphs, but since this is a nontrivial effort, we’ll send a separate PR once it’s ready.

File: python/sglang/srt/layers/pooler.py (hunk 5 above)
Comment (fortunecookiee) [reply python/sglang/srt/layers/pooler.py]: rebased.

**Review by kpham-sgl — APPROVED** [review]
LGTM. Thanks for addressing all of my comments! @fortunecookiee

**Review by sundar24295s — APPROVED** [review]

**Review by hnyls2002 — APPROVED** [review]
